### ISSUE TITLE
[YouTube Change] Throttling at 1MB/s

### ISSUE THREAD

**Issue by TheFrenchGhosty** [yt-dlp/yt-dlp#985]
## Description

YouTube seems to have implemented a new throttling method, an IP will get flagged and throttled consistently (and from the look of it, permanently) at ~1MB/s.

The test bellow is this video: https://www.youtube.com/watch?v=p2vOKrHzqcg downloaded with my [Watch on PC Script](https://github.com/TheFrenchGhosty/TheFrenchGhostys-Ultimate-YouTube-DL-Scripts-Collection/blob/master/scripts/Watch%20Scripts/Watch%20on%20PC%20Script/Watch%20on%20PC.sh) slightly modified, on a server with a 1Gbit/s connection.

Bumping the `--throttled-rate` to 1500 doesn't change anything, it always extract something that will get downloaded at ~1MB/s

If I run the same script, on the same server, but in a VM that is using a VPN it downloads at a normal speed.

I asked @pukkandan and he seems to agrees that it looks like throttling.

Note: an Invidious instance is also running on the IP, albeit, not public (only semi-private with a dozen of users) it's an IP that is used more than a "normal" end user IP.

## Verbose log

```
[debug] Command-line config: ['--throttled-rate', '100K', '--format', '(bestvideo[vcodec^=av01][height>=4320][fps>30]/bestvideo[vcodec^=vp9.2][height>=4320][fps>30]/bestvideo[vcodec^=vp9][height>=4320][fps>30]/bestvideo[vcodec^=avc1][height>=4320][fps>30]/bestvideo[height>=4320][fps>30]/bestvideo[vcodec^=av01][height>=4320]/bestvideo[vcodec^=vp9.2][height>=4320]/bestvideo[vcodec^=vp9][height>=4320]/bestvideo[vcodec^=avc1][height>=4320]/bestvideo[height>=4320]/bestvideo[vcodec^=av01][height>=2880][fps>30]/bestvideo[vcodec^=vp9.2][height>=2880][fps>30]/bestvideo[vcodec^=vp9][height>=2880][fps>30]/bestvideo[vcodec^=avc1][height>=2880][fps>30]/bestvideo[height>=2880][fps>30]/bestvideo[vcodec^=av01][height>=2880]/bestvideo[vcodec^=vp9.2][height>=2880]/bestvideo[vcodec^=vp9][height>=2880]/bestvideo[vcodec^=avc1][height>=2880]/bestvideo[height>=2880]/bestvideo[vcodec^=av01][height>=2160][fps>30]/bestvideo[vcodec^=vp9.2][height>=2160][fps>30]/bestvideo[vcodec^=vp9][height>=2160][fps>30]/bestvideo[vcodec^=avc1][height>=2160][fps>30]/bestvideo[height>=2160][fps>30]/bestvideo[vcodec^=av01][height>=2160]/bestvideo[vcodec^=vp9.2][height>=2160]/bestvideo[vcodec^=vp9][height>=2160]/bestvideo[vcodec^=avc1][height>=2160]/bestvideo[height>=2160]/bestvideo[vcodec^=av01][height>=1440][fps>30]/bestvideo[vcodec^=vp9.2][height>=1440][fps>30]/bestvideo[vcodec^=vp9][height>=1440][fps>30]/bestvideo[vcodec^=avc1][height>=1440][fps>30]/bestvideo[height>=1440][fps>30]/bestvideo[vcodec^=av01][height>=1440]/bestvideo[vcodec^=vp9.2][height>=1440]/bestvideo[vcodec^=vp9][height>=1440]/bestvideo[vcodec^=avc1][height>=1440]/bestvideo[height>=1440]/bestvideo[vcodec^=av01][height>=1080][fps>30]/bestvideo[vcodec^=vp9.2][height>=1080][fps>30]/bestvideo[vcodec^=vp9][height>=1080][fps>30]/bestvideo[vcodec^=avc1][height>=1080][fps>30]/bestvideo[height>=1080][fps>30]/bestvideo[vcodec^=av01][height>=1080]/bestvideo[vcodec^=vp9.2][height>=1080]/bestvideo[vcodec^=vp9][height>=1080]/bestvideo[vcodec^=avc1][height>=1080]/bestvideo[height>=1080]/bestvideo[vcodec^=av01][height>=720][fps>30]/bestvideo[vcodec^=vp9.2][height>=720][fps>30]/bestvideo[vcodec^=vp9][height>=720][fps>30]/bestvideo[vcodec^=avc1][height>=720][fps>30]/bestvideo[height>=720][fps>30]/bestvideo[vcodec^=av01][height>=720]/bestvideo[vcodec^=vp9.2][height>=720]/bestvideo[vcodec^=vp9][height>=720]/bestvideo[vcodec^=avc1][height>=720]/bestvideo[height>=720]/bestvideo[vcodec^=av01][height>=480][fps>30]/bestvideo[vcodec^=vp9.2][height>=480][fps>30]/bestvideo[vcodec^=vp9][height>=480][fps>30]/bestvideo[vcodec^=avc1][height>=480][fps>30]/bestvideo[height>=480][fps>30]/bestvideo[vcodec^=av01][height>=480]/bestvideo[vcodec^=vp9.2][height>=480]/bestvideo[vcodec^=vp9][height>=480]/bestvideo[vcodec^=avc1][height>=480]/bestvideo[height>=480]/bestvideo[vcodec^=av01][height>=360][fps>30]/bestvideo[vcodec^=vp9.2][height>=360][fps>30]/bestvideo[vcodec^=vp9][height>=360][fps>30]/bestvideo[vcodec^=avc1][height>=360][fps>30]/bestvideo[height>=360][fps>30]/bestvideo[vcodec^=av01][height>=360]/bestvideo[vcodec^=vp9.2][height>=360]/bestvideo[vcodec^=vp9][height>=360]/bestvideo[vcodec^=avc1][height>=360]/bestvideo[height>=360]/bestvideo[vcodec^=avc1][height>=240][fps>30]/bestvideo[vcodec^=av01][height>=240][fps>30]/bestvideo[vcodec^=vp9.2][height>=240][fps>30]/bestvideo[vcodec^=vp9][height>=240][fps>30]/bestvideo[height>=240][fps>30]/bestvideo[vcodec^=avc1][height>=240]/bestvideo[vcodec^=av01][height>=240]/bestvideo[vcodec^=vp9.2][height>=240]/bestvideo[vcodec^=vp9][height>=240]/bestvideo[height>=240]/bestvideo[vcodec^=avc1][height>=144][fps>30]/bestvideo[vcodec^=av01][height>=144][fps>30]/bestvideo[vcodec^=vp9.2][height>=144][fps>30]/bestvideo[vcodec^=vp9][height>=144][fps>30]/bestvideo[height>=144][fps>30]/bestvideo[vcodec^=avc1][height>=144]/bestvideo[vcodec^=av01][height>=144]/bestvideo[vcodec^=vp9.2][height>=144]/bestvideo[vcodec^=vp9][height>=144]/bestvideo[height>=144]/bestvideo)+(bestaudio[acodec^=opus]/bestaudio)/best', '--verbose', '--force-ipv4', '--ignore-errors', '--no-continue', '--no-overwrites', '--download-archive', 'archive.log', '--add-metadata', '--all-subs', '--embed-subs', '--output', '%(uploader)s - %(upload_date)s - %(title)s [%(id)s].%(ext)s', '--merge-output-format', 'mkv', '--batch-file', 'Source - Watch on PC.txt']
[debug] Batch file urls: ['https://www.youtube.com/watch?v=p2vOKrHzqcg']
[debug] Loading archive file 'archive.log'

[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] yt-dlp version 2021.09.02 (zip)
[debug] Python version 3.9.6 (CPython 64bit) - Linux-[REDACTED]
[debug] exe versions: ffmpeg 4.4, ffprobe 4.4, rtmpdump 2.4
[debug] Optional libraries: mutagen, pycryptodome, sqlite, websockets
[debug] Proxy map: {}
[debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=p2vOKrHzqcg
[youtube] p2vOKrHzqcg: Downloading webpage
[youtube] p2vOKrHzqcg: Downloading android player API JSON
[debug] Sort order given by extractor: quality, res, fps, source, codec:vp9.2, lang
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, source, vcodec:vp9.2(10), acodec, lang, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, id
[debug] Testing thumbnail 41
[info] p2vOKrHzqcg: Downloading 1 format(s): 313+251
[debug] Invoking downloader on "https://r3---sn-4g5lznes.googlevideo.com/videoplayback?expire=1631756162&ei=IktCYYLTDLTBmwfol4_IDA&ip=[REDACTED]&id=o-AOeE0qYG8e_tJyePCiIVGxWNXiihYgeCypFNzA9v3z7t&itag=313&source=youtube&requiressl=yes&mh=cg&mm=31%2C29&mn=sn-4g5lznes%2Csn-4g5ednsy&ms=au%2Crdu&mv=m&mvi=3&pl=21&initcwndbps=355000&vprv=1&mime=video%2Fwebm&gir=yes&clen=450924474&dur=524.457&lmt=1631710765060421&mt=1631733894&fvip=3&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5516222&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRgIhANhsvJ-MuJhAfS9w2LZzXZenxsNM6qbPWnLM60oyL_4_AiEA8E4-dYlz6wkxf_uel8RtZ4LOQRar4U--xlI0-pJW4Cg%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRgIhAMvt-N7oHycbCc8xyMTa4MuIUwz_y6xCS2-mhfSQ5hm4AiEArqoIDW5ueVf2jbH_HK3UJ4FmypstNy8foaw8ZfLJF9s%3D"
[download] Destination: Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].f313.webm
[download] 100% of 430.04MiB in 06:3507MiB/s ETA 00:001nown ETA
[debug] Invoking downloader on "https://r3---sn-4g5lznes.googlevideo.com/videoplayback?expire=1631756162&ei=IktCYYLTDLTBmwfol4_IDA&ip=[REDACTED]&id=o-AOeE0qYG8e_tJyePCiIVGxWNXiihYgeCypFNzA9v3z7t&itag=251&source=youtube&requiressl=yes&mh=cg&mm=31%2C29&mn=sn-4g5lznes%2Csn-4g5ednsy&ms=au%2Crdu&mv=m&mvi=3&pl=21&initcwndbps=355000&vprv=1&mime=audio%2Fwebm&gir=yes&clen=8177826&dur=524.481&lmt=1631710769717891&mt=1631733894&fvip=3&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5511222&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRgIhAOr9icFiAbTeeRWu34rTdDRQCwC6xX9lO7GghyXzQd0qAiEA8qm-6EI0KMN_faBrcp930r9KLB2AOSwGxRhptY7uBbM%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRgIhAMvt-N7oHycbCc8xyMTa4MuIUwz_y6xCS2-mhfSQ5hm4AiEArqoIDW5ueVf2jbH_HK3UJ4FmypstNy8foaw8ZfLJF9s%3D"
[download] Destination: Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].f251.webm
[download] 100% of 7.80MiB in 00:0700MiB/s ETA 00:009nown ETA
[Merger] Merging formats into "Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i 'file:Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].f313.webm' -i 'file:Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].f251.webm' -c copy -map 0:v:0 -map 1:a:0 'file:Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].temp.mkv'
Deleting original file Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].f251.webm (pass -k to keep)
Deleting original file Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].f313.webm (pass -k to keep)
[EmbedSubtitle] There aren't any subtitles to embed
[Metadata] Adding metadata to "Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i 'file:Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].mkv' -i 'file:Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].meta' -map 0 -dn -c copy -map_metadata 1 -metadata 'title=iPhone 13 - What Went Wrong?' -metadata date=20210914 -metadata 'description=[REDACTED] -metadata 'artist=Linus Tech Tips' 'file:Linus Tech Tips - 20210914 - iPhone 13 - What Went Wrong [p2vOKrHzqcg].temp.mkv'
```

**Comment by TheFrenchGhosty** [comment]
Research done, with the help of @pukkandan 

- Throttling doesn't affect the youtube player, it's correctly using 30-50MB/s
- `--extractor-args youtube:player_client=web` doesn't change anything to the throttling: "so whatever it is, affects both web and android" (quote from @pukkandan )

**Comment by 89z** [comment]
> Something I forgot to mention, it's sometimes throttled at 10MB/s

In what world is that considered throttling? My connection is 64 mb/s (8 MB/s), so thats literally faster than I can even download anything.

**Comment by TheFrenchGhosty** [comment]
@89z In a world where 1Gbit/s (~120MB/s) is the minimum speed you can get on server.

So yeah, this is throttling.

**Comment by pukkandan** [comment]
He is getting much faster speeds in browser with the same format. So there is indeed something going on here

**Comment by 89z** [comment]
I mean this just screams of making a problem where there is none. Ive seen other threads talking about 100 KB/s throttling, yeah thats concerning. But this is silly.

**Comment by TheFrenchGhosty** [comment]
@89z 

The throttling is almost always at 1MB/s which is indeed slow. 10MB/s is acceptable, sure, but it's still throttling.

**Comment by 89z** [comment]
Love the people downvoting me with no comment. [Heres to you][1].

On a constructive note, its possible YouTube could be doing TLS fingerprinting. I know for a fact that its being used with `android.clients.google.com`, so its possible. If you can provide a "good" and "bad" fingerprint, I can check into it some more. Just make an HTTPS request like this:

~~~
GET / HTTP/1.1
Host: client.tlsfingerprint.io:8443
~~~

and reply back with the results.

[1]:https://www.youtube.com/watch?v=bB7fgX6H9_M

**Comment by Ashish0804** [comment]
Why do you just start drama and delete comments? 🤦‍♂️ 

Here's the comment which got downvote btw
```
I mean this just screams of making a problem where there is none. Ive seen other threads talking about 100 KB/s throttling, yeah thats concerning. But this is silly.
```

**Comment by 89z** [comment]
@Ashish0804 you are sad. I have a potential solution to the problem, which no one has mention yet. and all you care about is pointless drama. Get over yourself.

**Comment by Ashish0804** [comment]
Does the solution involve me begging you to make a PR and if I don't do so you threaten to not make it public while also gloating about your fork?

**Comment by 89z** [comment]
> 
> 
> Does the solution involve me begging you to make a PR and if I don't do so you threaten to not make it public while also gloating about your fork?

@Ashish0804 What the hell is wrong with you? I have a constructive and on topic suggestion, which might be a fix to the issue. But all you can focus on is off topic drama. Drop the clown act.

**Comment by Ashish0804** [comment]
Again deleting comments
Go find another repo to troll
```
@Ashish0804 you are sad. I have a potential solution to the problem, which no one has mention yet. and all you care about is pointless drama. Get over yourself.
```

**Comment by 89z** [comment]
Heres another one - copy it too!

**Comment by nicolaasjan** [comment]
I have no throttling here in The Netherlands.
Downloading at 8-9 MB/sec, which is my max.

Could it be that they only do this when a user has a very high bandwidth and the nearest server can't - or doesn't want to -   handle that?

It would be interesting to know if users from different locations experience the same.

**Comment by xaviermarchegay** [comment]
> I have no throttling here in The Netherlands.
> Downloading at 8-9 MB/sec, which is my max.
> 
> Could it be that they only do this when a user has a very high bandwidth and the nearest server can't - or doesn't want to - handle that?
> 
> It would be interesting to know if users from different locations experience the same.

I definitely experienced the throttling yesterday (a few minutes after the issue was opened) on virtually every video and this morning.

This afternoon though (I am in France), I did not experience any throttling at all.

**Comment by mufapu** [comment]
> I definitely experienced the throttling yesterday (a few minutes after the issue was opened) on virtually every video and this morning.
> 
> This afternoon though (I am in France), I did not experience any throttling at all.
> 

Same experience (Belgium). No more issues here.

**Comment by coletdjnz** [comment]
> On a constructive note, its possible YouTube could be doing TLS fingerprinting. I know for a fact that its being used with `android.clients.google.com`, so its possible. If you can provide a "good" and "bad" fingerprint, I can check into it some more. Just make an HTTPS request like this:
> 
> ```
> GET / HTTP/1.1
> Host: client.tlsfingerprint.io:8443
> ```
> 
> and reply back with the results.

Interesting find. 

I'm not very knowledgeable in this area, but I will agree this could be a possibility too. I have seen what seemed to be some sort of fingerprinting when it comes to rate limiting/429 back in the day with the `get_video_info` endpoint.

**Comment by coletdjnz** [comment]
I downloaded a few channels in parallel to see if I could reproduce this yesterday, but didn't see anything amiss

**Comment by TheFrenchGhosty** [comment]
So... hum...

The throttling is gone since today.

I have been downloading correctly at 30MB/s all day now... Really no idea why... Nothing has changed.

**Comment by tyl0413** [comment]
> So... hum...
> 
> The throttling is gone since today.
> 
> I have been downloading correctly at 30MB/s all day now... Really no idea why... Nothing has changed.

I still have this issue.

**Comment by h1z1** [comment]
Try this - start a download, stop it and restart.  If it works then it's a proxy you're hitting.  Don't know if it's YT or carriers but I've run into several that throttle at the proxy (content servers on their network not yours).   

  Could also try grabbing the url and download with curl or something.  If that works (try twice to eliminate the first case),  then they are indeed fingerprinting.

**Comment by TheFrenchGhosty** [comment]
@h1z1 We already did both of those.

This is really throttling.

**Comment by h1z1** [comment]
@TheFrenchGhosty do you get the same cache for both attempts?

**Comment by TheFrenchGhosty** [comment]
@h1z1 https://github.com/yt-dlp/yt-dlp/issues/985#issuecomment-922388497

**Comment by TheFrenchGhosty** [comment]
1MB/s throttling is back today...

**Comment by coletdjnz** [comment]
does https://github.com/yt-dlp/yt-dlp/pull/1049 make any difference?

**Comment by sippeangelo** [comment]
I've found that they seem to be throttling me based on how far I've watched a video on the actual website, where the download would get throttled at 70% downloaded just because I had watched a video on the website to 70%. When I resumed it on the website the download also resumed at full speed...

**Comment by pukkandan** [comment]
Since this issue is bound to attract a lot of low quality comments, I'll put forth a few guidelines.
1. If you are unaffected by the throttling, cool, dont comment it. We are aware that only a very few people are affected at the time
2. If you are affected, but have nothing constructive to add, just give a thumbsup on the issue to let us know. Don't clutter the issue with "me too" comments
3. Don't ask "is this fixed?" etc. The issue will be closed when it is fixed

---



Due to the nature of the issue (only few people being affected), it is going to take help from the community to try and figure out a solution. Of all the "reported" cases, the only one that has currently been confirmed to be actual throttling and not a user error is TheFrenchGhosty's. So if you are affected, try these:

1. First confirm that it is in fact the same throttling issue and not something else. Unless you test these, we will treat it as a possibly false claim:
	* Confirm that you are in fact running the latest version (2021.10.10 at the time of writing). In these tests, don't use any custom config options, and always add `--no-config` to the command to make sure config files aren't loaded
	* If possible, try downloading with a VPN. Since throttling is IP bound, it should disappear. If it doesn't, there may be something else going on
	* See `-F` output and confirm the format note doesn't say `maybe throttled`. If it does, download another format
	* Confirm that the protocol is `https`. If it is not, the download is expected to be slow. You should use aria2c to speed it up (even if it is https, it may be useful to try with aria2c)
	* Open the video in browser and open stats for nerds. You can see the format_id (itag) and download speed there. Try downloading the same format as what your browser is giving. If it is a throttling issue, you'll get much lower speed in yt-dlp than in browser
	* Make sure you are not running in WSL. WSL is known to have an issue that reduces the download speed
	* At this point, we can be pretty sure you are being throttled. Now lets try to figure out what is going on
1. I have been told that youtube apps now use a chunk size of 2M instead of 10M. So try passing `--http-chunk-size`
1. If you have a VPN (which is not being throttled), try running `yt-dlp --get-url -f <format_id> <URL>` to get the URL through the VPN. Then turn VPN off and pass the URL you got into yt-dlp. Report back if you are still being throttled. This will help us determine whether the throttling is determined at the time of extraction, or at time of download
1. Play the video in browser and open developer tools. In the network tab, search for the `googlevideo.com` urls. Look for one with `itag=` the video format id (there will be many, just take any one). Copy the URL, remove any `&range=<number>-<number>` from the URL and pass it to yt-dlp. Report back if it is throttled. If it isn't, the browser is doing some magic to make sure it isnt throttled and we may have to go through youtube's javascripts again.
1. According to [this comment](https://github.com/yt-dlp/yt-dlp/issues/985#issuecomment-943575390), watching the video on the website fixes the throttling. It's honestly a long stretch, but might as well try it. Open the video in browser, and start download at the same time. When the download starts throttling (say at 30% of the video), scrub to around the same percent in the website and see if the throttling disappears

Will add more as new information comes in

PS: No, the `cant-reproduce` tag doesnt mean this is an invalid issue. It just means we don't know of a reliable way to reproduce the issue

**Comment by h1z1** [comment]
> The only one that has currently been confirmed to be actual throttling and not a user error is TheFrenchGhosty's. So if you are affected, try these:

I have confirmed what was reported.

> I have been told that youtube apps now use a chunk size of 2M instead of 10M. So try passing --http-chunk-size

I can confirm that as well.  My  _guess_ is their proxy's (or some edge providers), precache the first chunk.   One way around some types of throttling is to use a larger chunksize (anything with bursting for example).   My guess again is they don't like that or are interpreting it as an attack.

There may very well be different things happening or some common gear on their backend we'll never know about.   The throttling is very much real however.

Regarding VPN use it's also worth noting vpn providers tend to have better access (routing) to sites like YT so while it may appear YT is throttling some networks but not others, it could be those vpn ip blocks haven't been identified yet.   They can also turn their experiment on/off at will.

**Comment by Techjar** [comment]
1, 2, 3 and 5 don't make any difference whatsoever. With 4 I see some interesting behavior: the download is fast for the first second, then goes back to being slow. I can repeatedly stop and resume the download and it will again be fast for the first second, then slow, no matter what byte it resumes at.

**Comment by TheFrenchGhosty** [comment]
@pukkandan The throttling is back... and it even worse than before: it's basically impossible to download multiples videos in a row "quickly". Again, on a VM behind a VPN running on the machine, I get 30-40MB/s without any issue.

Here's some recording of what can happen:

Either, dips to 100-1000 KB/s for seconds:

https://user-images.githubusercontent.com/47571719/144116868-6f512e07-f29b-4cd6-8149-eab94f7a5a39.mp4

Or, and this is where it's where it's a real problem, full speed for X amount of seconds, then complete stop for 10+ seconds, then full speed (or 100-1000KB/s) for X amount of seconds (if things are "okay" the download will complete after X minutes, if not, the `--throttled-rate` will trigger... and it will restart the download from the start):


https://user-images.githubusercontent.com/47571719/144117182-1bc0ad37-2698-4760-ae8b-c8f934029262.mp4

To add to that, Invidious instances are now directly impacted by this: https://github.com/iv-org/invidious/issues/2407#issuecomment-981062559

That's... bad.

**Comment by TheFrenchGhosty** [comment]
Some news since it has been a while: the throttling is currently going on and off, depending on the days.

Most days it works at ~10MB/s (slower than before, but "okay" speed), and some day it's throttled bellow 1MB/s, I really don't know what's the logic behind it.

The full stop for X seconds isn't really a thing anymore (at least from what I now)... so I guess they're just doing tests...

YouTube is such a pain...

**Comment by dfgyuri** [comment]
Speeds are still throttled on certain vids at certain times on latest build to ~1MB/s. Ive found it applies to older/vids with lower views as it hits hard when I try to download a whole channel with old archived streams but the newer more viewed videos download at full speed.  I don't think ive ever been throttled trying to download new and popular vids, only older and low view count ones, especially if theyre long.

**Comment by MinePlayersPE** [comment]
@dfgyuri Can you show `-F` output for any of the low viewcount throttled videos?
> Confirm that the protocol is https. If it is not, the download is expected to be slow. You should use aria2c to speed it up (even if it is https, it may be useful to try with aria2c)
^ This is likely to happen on low viewcount videos

**Comment by dfgyuri** [comment]
@MinePlayersPE Non throttled speeds dl at around 30MB/s for me, this one was especially bad at only ~60KB/s
`mpv>youtube-dl.exe -F https://www.youtube.com/watch?v=U7koqKgd14g
[youtube] U7koqKgd14g: Downloading webpage
[youtube] U7koqKgd14g: Downloading android player API JSON
[youtube] U7koqKgd14g: Downloading MPD manifest
[youtube] U7koqKgd14g: Downloading MPD manifest
[info] Available formats for U7koqKgd14g:
ID       EXT   RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC        VBR ACODEC      ABR     ASR MORE INFO
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
sb2      mhtml 48x27          │                  mhtml │ images                                    storyboard
sb1      mhtml 80x45          │                  mhtml │ images                                    storyboard
sb0      mhtml 160x90         │                  mhtml │ images                                    storyboard
139-dash m4a                  │   40.90MiB   56k dash  │ audio only        mp4a.40.5   56k 22050Hz DASH audio, m4a_dash
139      m4a                  │   40.90MiB   48k https │ audio only        mp4a.40.5   48k 22050Hz low, m4a_dash
140-dash m4a                  │  108.55MiB  137k dash  │ audio only        mp4a.40.2  137k 44100Hz DASH audio, m4a_dash
140      m4a                  │  108.55MiB  129k https │ audio only        mp4a.40.2  129k 44100Hz medium, m4a_dash
251-dash webm                 │  117.71MiB  172k dash  │ audio only        opus       172k 48000Hz DASH audio, webm_dash
251      webm                 │  117.71MiB  140k https │ audio only        opus       140k 48000Hz medium, webm_dash
17       3gp   176x144      8 │   65.23MiB   77k https │ mp4v.20.3     77k mp4a.40.2    0k 22050Hz 144p
160      mp4   256x144     30 │ ~ 92.72MiB  108k dash  │ avc1.4d400b  108k video only              DASH video, mp4_dash
278      webm  256x144     30 │ ~ 81.56MiB   95k dash  │ vp9           95k video only              DASH video, webm_dash
133      mp4   426x240     30 │ ~207.76MiB  242k dash  │ avc1.4d400c  242k video only              DASH video, mp4_dash
242      webm  426x240     30 │ ~189.30MiB  220k dash  │ vp9          220k video only              DASH video, webm_dash
134-dash mp4   640x360     30 │  372.03MiB  622k dash  │ avc1.4d401e  622k video only              DASH video, mp4_dash
134      mp4   640x360     30 │  372.03MiB  443k https │ avc1.4d401e  443k video only              360p, mp4_dash
18       mp4   640x360     30 │  436.51MiB  520k https │ avc1.42001E  520k mp4a.40.2    0k 44100Hz 360p
243      webm  640x360     30 │ ~348.32MiB  405k dash  │ vp9          405k video only              DASH video, webm_dash
135      mp4   854x480     30 │ ~991.59MiB 1155k dash  │ avc1.4d4014 1155k video only              DASH video, mp4_dash
244      webm  854x480     30 │ ~646.16MiB  752k dash  │ vp9          752k video only              DASH video, webm_dash
136      mp4   1280x720    30 │ ~  1.94GiB 2310k dash  │ avc1.4d400c 2310k video only              DASH video, mp4_dash
22       mp4   1280x720    30 │ ~  1.31GiB 1565k https │ avc1.64001F 1565k mp4a.40.2    0k 44100Hz 720p
247      webm  1280x720    30 │ ~  1.26GiB 1505k dash  │ vp9         1505k video only              DASH video, webm_dash
298-dash mp4   1280x720    60 │    1.64GiB 3018k dash  │ avc1.640020 3018k video only              DASH video, mp4_dash
298      mp4   1280x720    60 │    1.64GiB 2005k https │ avc1.640020 2005k video only              720p60, mp4_dash
302      webm  1280x720    60 │ ~  2.22GiB 2646k dash  │ vp9         2646k video only              DASH video, webm_dash`

**Comment by MinePlayersPE** [comment]
> 1280x720    60 │ ~  2.22GiB 2646k **dash**  │ vp9         2646k video only              DASH video, webm_dash

**Comment by soredake** [comment]
Getting the same problem, adding `--extractor-args youtube:player_client=web` helped with speed for now.

**Comment by soredake** [comment]
"web" user-agent stopped giving me decent speed on throttled videos, switch to "ios" user-agent working, for now, again.

**Comment by soredake** [comment]
So far, after 21 days, i don't have problems with speed with `ios` user-agent.

**Comment by aliafshany** [comment]
hi there @soredake 
how did you mange to change to ios user-agent?

best

**Comment by soredake** [comment]
> hi there @soredake 
> how did you mange to change to ios user-agent?
> 
> best

`ytdl-raw-options=extractor-args="youtube:player_client=ios"` to mpv.conf

**Comment by soredake** [comment]
Using `android` player client is also fine, you are limited to vp9/opus and can play 4k vs h264/aac/1080p on `ios` player client.

**Comment by soredake** [comment]
If use `android,web` (which is default) or just `web` i frequently get throttling or video (only video, sound is fine) doesn't even load saying this:
```
Could not read init fragment. 
EDL: source file has unknown duration.
```

**Comment by pukkandan** [comment]
> If use `android,web` (which is default) or just `web` i frequently get throttling or video (only video, sound is fine) doesn't even load saying this:
> 
> ```
> Could not read init fragment. 
> EDL: source file has unknown duration.
> ```
> 
> full log when video refuse to load: [youtube-blackscreen-error.txt](https://github.com/yt-dlp/yt-dlp/files/8884433/youtube-blackscreen-error.txt)

Bruh, are all your previous comments based on using `mpvnet`? The download here is not being done through yt-dlp. Unless you can reproduce the issue directly in yt-dlp, open issue in the relevent trackers

**Comment by soredake** [comment]
@pukkandan why does this matter if i use mpvnet or not? yt-dlp gives me links that cannot be opened or being trottled if i use `web` useragent, but all is fine when i use `ios` or `android`.

**Comment by pukkandan** [comment]
> @pukkandan why does this matter if i use mpvnet or not? yt-dlp gives me links that cannot be opened or being trottled if i use web useragent, but all is fine when i use ios or android.

Because the URL is not the only thing that matters. It matters exactly **how** the data is fetched for download. yt-dlp does various "tricks" in the download process to avoid throttling (My guess is that your issue is caused due to using an "http chunk size" > 10M, which YouTube is known to throttle for `WEB` URLs). If the link throttles when used externally, the external downloader must implement those tricks - no use discussing it here

**Comment by EndlessEden** [comment]
Can confirm @TheFrenchGhosty initial post. When a video is open in the browser (unloaded, not buffering and not playing), yt-dlp is throttled to 300-600KiB/s. 
The moment the video page is loaded in a browser, yt-dlp's rate will begin to hit the bandwidth caps of the host system. If the page is closed, it immediately returns to previous cap. Its reproducible. (Examples: "https://www.youtube.com/watch?v=Rej7aTENLKY" "https://www.youtube.com/watch?v=MHyJ1J8vJaY")

(If this doesnt happen on VPN's, its probably due to shared address spacing resulting higher likelihood of the above behavior occurring on another host connected to the same VPN provider. Or Alternatively, the VPN provider operating their own CDN that doesnt comply with YT cache constraints[ex: lancache].)

However, this only applies for "uploaded" content. Content that's streaming or was previously streamed (EG: not cached by a local CDN), its rate-limit capped at 600KiB/s regardless (its not in the browser and peaks at 3.2MiB/s). I would blame FFMPEG for rate-capping for during streams, however, post-stream where the fragments are downloaded directly is still just as slow. Leading to suggest some other behavior occurs now. (Example: https://www.youtube.com/watch?v=amWa76ym0_o)

---

YT must be using a script, that informs the CDN network to uncap player rate and has a keep-alive.

**Comment by soredake** [comment]
`--extractor-args youtube:player_client=android` helps me with speed for now.

**Comment by dextertd** [comment]
> `--extractor-args youtube:player_client=android` helps me with speed for now.

This made my download speeds very volatile (sudden bursts then going back to sub 1MB/s) ­– but on the whole *I think* improved the average.

However, I coincidentally found a much better solution. Playing *any* youtube video brought my speeds back up to the 3-8MB/s I expected. So I just opened a private browsing window to the homepage, clicked a video, muted the tab, and let a 2 hour video run. It autoplayed to another video until I tabbed back to terminal and left it running for the night

The next morning still saw good speeds long after the video had ended. I had left that `--extractor-args` on the whole time.

This is slightly different to the [previous advice](https://github.com/yt-dlp/yt-dlp/issues/985#issuecomment-943575390) of playing the specific video you are attempting to download; this is just playing any video on youtube.

**Comment by slagjaw** [comment]
> Or, and this is where it's where it's a real problem, full speed for X amount of seconds, then complete stop for 10+ seconds, then full speed (or 100-1000KB/s) for X amount of seconds

I'm running into a similar issue that no matter how many retries, will temporarily start near my host cap and then slowly bleed to a number that's below (never exceeding) 1MiB/s; which sometimes, but not always, stops completely for X amount of time and resumes. The only way to circumnavigate this is by retrying and the cycle repeats.

> --extractor-args youtube:player_client=android helps me with speed for now.

I did some testing and noticed that _without_ the android client the numbers were basically halved. Still, while enabled, there is no significant boost and it remains below 1MiB/s given enough time.

Cannot confirm the browser method that @EndlessEden explains.

**Comment by 0x70b1a5** [comment]
> --extractor-args youtube:player_client=android helps me with speed for now.

Didn't work for me.

> Playing any youtube video brought my speeds back up to the 3-8MB/s I expected. 

Also didn't work for me.

Going to be a real bummer if we end up in a stealth war with Google over this.

**Comment by nicolaasjan** [comment]
@0x70b1a5 
The current throttling is [another issue](https://github.com/yt-dlp/yt-dlp/issues/6369).
Will probably be fixed in a [soon to be expected release](https://github.com/yt-dlp/yt-dlp/issues/6369#issuecomment-1450068901). 🙂

**Comment by pukkandan** [comment]
Outdated
